tests/render/renderer/test_notebook_renderer.py

```diff
@@ -1495,3 +1500,76 @@ def test_complex_suite(warning_suite):
     for obs_cell, expected_cell in zip(obs["cells"], expected["cells"]):
         assert obs_cell == expected_cell
     assert obs == expected
+
+
+def test_notebook_execution_with_pandas_backend(titanic_data_context):
+    """
+    To set this test up we:
+
+    - create a suite
+    - add a few expectations (both table and column level)
+    - verify that no validations have happened
+    - create the suite edit notebook by hijacking the private cli method
+
+
+    We then:
+    - execute that notebook (Note this will raise various errors like
+    CellExecutionError if any cell in the notebook fails
+    - create a new context from disk
+    - a validation has been run with our expectation suite
+    """
+    context = titanic_data_context
+    root_dir = context.root_directory
+    uncommitted_dir = os.path.join(root_dir, "uncommitted")
+    suite_name = "warning"
+
+    context.create_expectation_suite(suite_name)
+    csv_path = os.path.join(root_dir, "..", "data", "Titanic.csv")
+    batch_kwargs = {"datasource": "mydatasource", "path": csv_path}
+    batch = context.get_batch(batch_kwargs, suite_name)
+    batch.expect_table_column_count_to_equal(1)
+    batch.expect_table_row_count_to_equal(1313)
+    batch.expect_column_values_to_be_in_set("Sex", ["female", "male"])
+    batch.save_expectation_suite(discard_failed_expectations=False)
+
+    # Sanity check test setup
+    suite = context.get_expectation_suite(suite_name)
+    assert len(suite.expectations) == 3
+    assert context.list_expectation_suite_names() == [suite_name]
+    assert context.list_datasources() == [
+        {"class_name": "PandasDatasource", "name": "mydatasource"}
+    ]
+    assert context.get_validation_result("warning") == {}
+
+    # Create notebook
+    json_batch_kwargs = json.dumps(batch_kwargs)
+    _suite_edit(
+        suite_name,
+        "mydatasource",
+        directory=root_dir,
+        jupyter=False,
+        batch_kwargs=json_batch_kwargs,
+    )
+    edit_notebook_path = os.path.join(uncommitted_dir, "warning.ipynb")
+    assert os.path.isfile(edit_notebook_path)
+
+    with open(edit_notebook_path, "r") as f:
+        nb = nbformat.read(f, as_version=4)
+
+    # Run notebook
+    ep = ExecutePreprocessor(timeout=600, kernel_name="python3")
+    ep.preprocess(nb, {"metadata": {"path": uncommitted_dir}})
+
+    # Useful to inspect executed notebook
+    with open(os.path.join(uncommitted_dir, "output.ipynb"), "w") as f:
+        nbformat.write(nb, f)
+
+    # Assertions about output
+    context = DataContext(root_dir)
+    obs_validation_result = context.get_validation_result("warning")
+    assert obs_validation_result.statistics == {
```

Comment:
Could we assert that the new expectation suite is the same as the old expectation suite?

Reply:
ooh yes nice

Reply:
Done